src/fiber_prompt/wasmfx/wasmfx_impl.c

```diff
@@ -0,0 +1,181 @@
+    uint32_t fiber_index, void* arg, fiber_result_t* result);
+
+extern import("wasmfx_suspend_to") void* wasmfx_suspend_to(void* arg);
+
+// this should grow both the table and name
+static cont_table_index_t wasmfx_acquire_table_index(void) {
+  uintptr_t table_index;
+  if (cont_table_unused_size > 0) {
+    // There is an entry in the continuation table that has not been used so
+    // far.
+    table_index = cont_table_capacity - cont_table_unused_size;
+    cont_table_unused_size--;
+#ifdef FIBER_WASMFX_PRESERVE_SHADOW_STACK
+    init_shadow_stack(table_index);
+#endif
+  } else if (free_list_size > 0) {
+    // We can pop an element from the free list stack.
+    table_index = free_list[free_list_size - 1];
+    free_list_size--;
+  } else {
+    // We have run out of table entries.
+    uint32_t new_cont_table_capacity = 2 * cont_table_capacity;
+
+    // Ask wasm to grow the table by the previous size, and we grow the
+    // `free_list` ourselves.
+    wasmfx_grow_tables(cont_table_capacity);
+    free(free_list);
+    free_list = malloc(sizeof(uint32_t) * new_cont_table_capacity);
+
+    // We added `cont_table_capacity` new entries to the table, and then
+    // immediately consume one for the new continuation.
+    cont_table_unused_size = cont_table_capacity - 1;
+    table_index = cont_table_capacity;
+    cont_table_capacity = new_cont_table_capacity;
+
+#ifdef FIBER_WASMFX_PRESERVE_SHADOW_STACK
+    sstack_bottom_ptrs =
+        realloc(sstack_bottom_ptrs, sizeof(void*) * new_cont_table_capacity);
+    sstack_current_ptrs =
+        realloc(sstack_current_ptrs, sizeof(void*) * new_cont_table_capacity);
+    init_shadow_stack(table_index);
+#endif
+  }
+  return table_index;
+}
+
+static void wasmfx_release_table_index(cont_table_index_t table_index) {
+  free_list[free_list_size] = table_index;
+  free_list_size++;
+#ifdef FIBER_WASMFX_PRESERVE_SHADOW_STACK
+  // Note that we do not free shadow stacks, but simply reset them for
+  // subsequent re-use.
+  void* shadow_stack_usable_top =
+      ((char*)sstack_bottom_ptrs[table_index]) + cont_shadow_stack_size - 0x10;
+  sstack_current_ptrs[table_index] = shadow_stack_usable_top;
+#endif
+}
+
+fiber_t fiber_alloc(fiber_entry_point_t entry) {
+  cont_table_index_t table_index = wasmfx_acquire_table_index();
+  wasmfx_indexed_cont_new(entry, table_index);
+
+  return (fiber_t)table_index;
+}
+
+void fiber_free(fiber_t fiber) {
+  cont_table_index_t table_index = (cont_table_index_t)fiber;
+
+  // NOTE: Currently, fiber stacks are deallocated only when the continuation
+  // returns. Thus, the only thing we can do here is releasing the table index.
+  wasmfx_release_table_index(table_index);
+}
+
+void* fiber_resume_with(fiber_t fiber, void* arg, fiber_result_t* result) {
+  cont_table_index_t table_index = (cont_table_index_t)fiber;
+  return wasmfx_indexed_resume_with(table_index, arg, result);
+}
+
+void* fiber_yield_to(void* arg) {
```

Comment:
Definitively need a prompt here, even if it isn't used. E.g.
```suggestion
void* fiber_yield_to(__attribute__((unused)) prompt_t p, void* arg) {
```